Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -312,22 +316,27 @@ end ProdAdicCompletions
 open ProdAdicCompletions.IsFiniteAdele
 
 /-- The finite adèle ring of `R` is the restricted product over all maximal ideals `v` of `R`
-of `adicCompletion` with respect to `adicCompletionIntegers`. -/
-def FiniteAdeleRing : Type _ := (
-  { carrier := {x : K_hat R K | x.IsFiniteAdele}
-    mul_mem' := mul
-    one_mem' := one
-    add_mem' := add
-    zero_mem' := zero
-    algebraMap_mem' := algebraMap'
-  } : Subalgebra K (K_hat R K))
+of `adicCompletion`, with respect to `adicCompletionIntegers`. -/
+def FiniteAdeleRing : Type _ := {x : K_hat R K // x.IsFiniteAdele}
```

Comment:
Why do you prefer to define this as a subtype, rather than a subalgebra?

Reply:
If you look at the original version of this PR [here](https://github.com/leanprover-community/mathlib4/pull/13705/commits/6a5c0f781c81f21d237dc0fed160ef38be1ab868#diff-e3f60199ddcf6fe696a3eed7602eded09cf575d804ab32af60be92b53c2cabb6R339-R347) you'll see what happens if you make it a coerced subalgebra; the `simp` lemma you need to get things back under control is horrible. Now we make it as a subtype and get the algebra structure via the subalgebra.

Reply:
I think María Inés was asking why this isn't `def FiniteAdeleRing : Subalgebra K (K_hat R K)`. To quote Kevin from above:
> The problem with that is that `K_hat R K`, the thing it's a subalgebra of, is a technical intermediate thing, but `FiniteAdeleRing R K` is the fundamental definition and I think it deserves to be a type. There's a problem with having coerced terms as fundamental objects, and that's that typeclass inference goes crazy trying to deal with it. Here's the issue. We need to let `FiniteAdeleRing` be a type because it has, for example, a topology on it (and it's not the subspace topology coming from K_hat). So if it's a subalgebra then the code will be full of `↥FiniteAdeleRing`, and when typeclass sees that it wants e.g. a topological space structure on this, or a coercion from this to somewhere else, it is quick to start looking looking at the auxiliary object `K_hat` for answers. This leads to timeouts (we saw it with `↥(𝓞 K)` in the theory of number fields and people were complaining about timeouts for about a year) and this is exactly why we changed the definition of ring of integers from a term to a type (and it solved the problems!)

We should add a library note for this.